Disable MarkerSortUtil batch sorting. (closes #53)

jukzi opened the pull request:
Can be reenabled with org.eclipse.ui.MarkerSortUtil.batchSize=10000

Batch sorting was intended to decrease memory consumption during sorting
Marker View. But it is very slow especially when CollationKeys are
calculated. Today temporary memory consumption is not as much an issue
any more, but performance is.

Performance difference is visible with MarkerSortUtilTest especially
when increasing ARRAYSIZE to a million (22sec vs 2sec) - even without
CollationKeys.

jukzi committed:
Disable MarkerSortUtil batch sorting. (Fixes #53)

Can be reenabled with org.eclipse.ui.MarkerSortUtil.batchSize=10000

Batch sorting was intended to decrease memory consumption during sorting
Marker View. But it is very slow especially when CollationKeys are
calculated. Today temporary memory consumption is not as much an issue
any more, but performance is. Also test showed that the memory benefit
is not archived with batching.

Performance difference is visible with MarkerSortUtilTest especially
when increasing ARRAYSIZE to a million (22sec vs 2sec) - even without
CollationKeys.

jukzi commented:
@iloveeclipse could you please verify your workspace with many markers does not make trouble during sorting markers view.

iloveeclipse commented:
I would recommend just create simple project with 10000 files and create 10 different markers per file. You do not need anything special here. You can use my project generator to do that. Just break generated java code and that's it. 
https://github.com/iloveeclipse/java-project-generator

jukzi committed:
Cache CollationKey during Sorting (Fixes #53)

Collation Key computation is pretty time consuming. Since the messages
are typically constant the calculated CollationKeys can typically be
reused many times during sorting.

jukzi commented:
with the generated project from https://github.com/iloveeclipse/java-project-generator/pull/1 the problem is pretty obvious, i even get a waiting popup during sort.
With both commits the perfromance is not only ~ 4sec vs instant sort but also with LESS memory allocation during sort (i do not know why). Both commits are needed to archive the instant sort.

jukzi commented:
![image](https://user-images.githubusercontent.com/51790620/166825598-3d5a6365-106d-41c6-98e4-46dfd0aaa6a3.png)

mickaelistria commented:
This looks good to me.
While looking a bit at this code, wouldn't it be worth trying to refactor it so it becomes some "SortWithProgrssUtil" with generified methods as I don't see any logic here that's specific to markers.

iloveeclipse commented on bundles/org.eclipse.ui.ide/src/org/eclipse/ui/internal/views/markers/MarkerEntry.java:
```diff
@@ -385,6 +388,7 @@ Map<String, Object> getCache() {
 	@Override
 	void clearCache() {
```
`clearCache()` seem to be called on **every** marker. Does it make sense to have a dedicated method for the static cache cleanup? Just to avoid the map is recreated 100_000 times, also it sounds strange to cleanup static data in non static context.

jukzi replied:
> `clearCache()` seem to be called on **every** marker

done

jukzi committed:
static MarkerEntry.clearCollationCache()

to only clear collationCache once.

Change-Id: Ife1b4f00192663e20dd6163e51df949707ede4d3

jukzi commented:
> This looks good to me.
> While looking a bit at this code, wouldn't it be worth trying to refactor it so it becomes some "SortWithProgrssUtil" with generified methods as I don't see any logic here that's specific to markers.

Actually the cancel is the only Goodie we loose by disabling the batched sort. But i prefer to sort in 1 sec then being able to cancel a longer running sort.
The only marker specific is the clear of caches.

iloveeclipse commented on bundles/org.eclipse.ui.ide/src/org/eclipse/ui/internal/views/markers/MarkerEntry.java:
```diff
@@ -219,7 +221,8 @@ CollationKey getCollationKey(String attribute, String defaultValue) {
 		}
-		CollationKey key = Collator.getInstance().getCollationKey(attributeValue);
+		CollationKey key = collationCache.computeIfAbsent(attributeValue,
+				k -> Collator.getInstance().getCollationKey(attributeValue));
```
I haven't tried that in the debugger yet, but what kind of values are cached here, and why is this cache **per marker**? Shouldn't that be cache per view, or static one?

jukzi replied:
ancient code.. currently it's only the IMarker.MESSAGE of each marker

jukzi commented:
any objections to merge?
